File: artifacts/api-server/src/agents/shared/validation.ts
/**
 * SQL validators used by all agent tools that execute generated SQL.
 *
 * Defence in depth: the system prompt tells the model what's allowed, this
 * runtime check enforces it. Per-agent tools wrap these validators around
 * every model-generated SQL string BEFORE it reaches pg.
 */

const SELECT_ONLY = /^\s*(WITH\s|SELECT\s)/i;
const FORBIDDEN = /\b(INSERT|UPDATE|DELETE|DROP|ALTER|TRUNCATE|GRANT|REVOKE|CREATE|MERGE|COPY|VACUUM|ANALYZE)\b/i;

export function assertSelectOnly(sql: string): void {
  if (!SELECT_ONLY.test(sql)) {
    throw new Error("Only SELECT (or WITH ... SELECT) statements are allowed.");
  }
  if (FORBIDDEN.test(sql)) {
    throw new Error("SQL contains a forbidden DDL/DML keyword. Only SELECT is allowed.");
  }
}

/**
 * Verifies that the SQL only references tables in the allowed schemas. This
 * is a heuristic (looks for `"schema"."table"` and `schema.table` patterns);
 * the canonical defence is running the query under a Postgres role that has
 * SELECT permission only on the project's schemas, which the per-agent runner
 * does via SET search_path + a least-privilege role (see deferred work).
 */
export function assertSchemaScope(sql: string, allowedSchemas: ReadonlyArray<string>): void {
  const allowed = new Set(allowedSchemas.map((s) => s.toLowerCase()));
  const matches = sql.matchAll(/(?:"([a-z0-9_]+)"|([a-z0-9_]+))\s*\.\s*(?:"[a-z0-9_]+"|[a-z0-9_]+)/gi);
  for (const m of matches) {
    const schema = (m[1] ?? m[2] ?? "").toLowerCase();
    if (schema && !allowed.has(schema)) {
      throw new Error(`SQL references schema "${schema}" which is not allowed for this project. Allowed: ${[...allowed].join(", ")}.`);
    }
  }
}

/**
 * For DDL-emitting transformation SQL, allow CREATE OR REPLACE VIEW and
 * CREATE TABLE ... AS SELECT scoped to a single allowed warehouse schema.
 * Used by DataEngineerAgent.apply_transformation.
 */
const TRANSFORM_ALLOWED = /^\s*CREATE\s+(OR\s+REPLACE\s+)?(VIEW|TABLE|MATERIALIZED\s+VIEW)\s+/i;

export type TransformationDdlKind = "table" | "view";

/**
 * Maps the agent's `kind` field to the DDL kind we expect:
 *   cleanse / join / rename            → physical TABLE
 *   aggregate / filter / view          → VIEW (always live)
 * Anything else falls back to VIEW (safer — no materialisation surprise).
 */
export function ddlKindForTransformation(kind: string): TransformationDdlKind {
  const k = kind.toLowerCase();
  if (k === "cleanse" || k === "join" || k === "rename") return "table";
  return "view";
}

export function assertTransformationSql(
  sql: string,
  warehouseSchema: string,
  expectedDdl?: TransformationDdlKind,
): void {
  if (!TRANSFORM_ALLOWED.test(sql)) {
    throw new Error("Transformation SQL must start with CREATE [OR REPLACE] VIEW / TABLE / MATERIALIZED VIEW.");
  }
  if (/\bDROP\b|\bDELETE\b|\bTRUNCATE\b|\bGRANT\b|\bREVOKE\b|\bALTER\b/i.test(sql)) {
    throw new Error("Transformation SQL contains a forbidden destructive keyword.");
  }
  // First quoted identifier after CREATE ... must be the target warehouse schema.
  // The schema name is now per-project (proj_{id}_warehouse), so we match the
  // longer identifier pattern instead of the literal "warehouse".
  const head = sql.match(/CREATE\s+(OR\s+REPLACE\s+)?(VIEW|TABLE|MATERIALIZED\s+VIEW)\s+"?([a-z0-9_]+)"?\s*\.\s*"?[a-z0-9_]+"?/i);
  if (!head) {
    throw new Error(`Transformation SQL must use a fully-qualified target like "${warehouseSchema}"."my_view".`);
  }
  const ddl = head[2].toUpperCase().startsWith("VIEW") || head[2].toUpperCase().includes("MATERIALIZED")
    ? "view"
    : "table";
  if (head[3].toLowerCase() !== warehouseSchema.toLowerCase()) {
    throw new Error(`Transformation must target schema "${warehouseSchema}", got "${head[3]}".`);
  }
  if (expectedDdl && ddl !== expectedDdl) {
    throw new Error(
      `This transformation should produce a ${expectedDdl.toUpperCase()} (per its kind), but the SQL creates a ${ddl.toUpperCase()}.`,
    );
  }
}

/**
 * Rewrites the head of a CREATE statement so the DDL kind matches what the
 * transformation `kind` says it should be. Used on apply to make older
 * proposals still runnable after the rule change — e.g. an aggregate that
 * was originally written as CREATE TABLE gets normalised to CREATE OR REPLACE VIEW.
 *
 * Returns the (possibly rewritten) SQL.
 */
export function normalizeTransformationDdl(
  sql: string,
  expectedDdl: TransformationDdlKind,
): string {
  const match = sql.match(/^(\s*)CREATE\s+(OR\s+REPLACE\s+)?(VIEW|TABLE|MATERIALIZED\s+VIEW)(\s+)/i);
  if (!match) return sql;
  const [whole, leading, , kind] = match;
  const currentDdl = kind.toUpperCase().includes("VIEW") ? "view" : "table";
  if (currentDdl === expectedDdl) return sql;

  const replacement = expectedDdl === "view"
    ? `${leading}CREATE OR REPLACE VIEW `
    : `${leading}CREATE TABLE `;

  let rewritten = sql.replace(whole, replacement);

  // CREATE TABLE needs an "AS" before the SELECT — if we're rewriting a VIEW
  // (which has its own "AS") to a TABLE, the AS is already there. Going the
  // other direction, "CREATE OR REPLACE VIEW ... AS SELECT" is also fine.
  // But "CREATE TABLE ... SELECT" (missing AS) needs fixing if the original
  // was a view-style statement. We detect that by checking for "AS\s+SELECT".
  if (!/\bAS\s+SELECT/i.test(rewritten) && /\bSELECT\b/i.test(rewritten)) {
    rewritten = rewritten.replace(/\)\s*SELECT/i, ") AS SELECT").replace(/"\s+SELECT/i, '" AS SELECT');
  }
  return rewritten;
}

File: services/agents-py/app/agents/shared/validation.py
"""SQL validators — defence-in-depth port of `src/agents/shared/validation.ts`
plus `assertMeasureFormula` from the metric-architect executor.

These run on every model-generated SQL string BEFORE it reaches Postgres.
"""
from __future__ import annotations

import re
from typing import Literal, Sequence

TransformationDdlKind = Literal["table", "view"]


class SqlValidationError(ValueError):
    """Raised when generated SQL fails a guardrail."""


_SELECT_ONLY = re.compile(r"^\s*(WITH\s|SELECT\s)", re.IGNORECASE)
_FORBIDDEN = re.compile(
    r"\b(INSERT|UPDATE|DELETE|DROP|ALTER|TRUNCATE|GRANT|REVOKE|CREATE|MERGE|COPY|VACUUM|ANALYZE)\b",
    re.IGNORECASE,
)


def assert_select_only(sql: str) -> None:
    if not _SELECT_ONLY.search(sql):
        raise SqlValidationError("Only SELECT (or WITH ... SELECT) statements are allowed.")
    if _FORBIDDEN.search(sql):
        raise SqlValidationError(
            "SQL contains a forbidden DDL/DML keyword. Only SELECT is allowed."
        )


_SCHEMA_REF = re.compile(
    r'(?:"([a-z0-9_]+)"|([a-z0-9_]+))\s*\.\s*(?:"[a-z0-9_]+"|[a-z0-9_]+)',
    re.IGNORECASE,
)


def assert_schema_scope(sql: str, allowed_schemas: Sequence[str]) -> None:
    allowed = {s.lower() for s in allowed_schemas}
    for m in _SCHEMA_REF.finditer(sql):
        schema = (m.group(1) or m.group(2) or "").lower()
        if schema and schema not in allowed:
            raise SqlValidationError(
                f'SQL references schema "{schema}" which is not allowed for this project. '
                f"Allowed: {', '.join(allowed)}."
            )


_TRANSFORM_ALLOWED = re.compile(
    r"^\s*CREATE\s+(OR\s+REPLACE\s+)?(VIEW|TABLE|MATERIALIZED\s+VIEW)\s+", re.IGNORECASE
)
_TRANSFORM_DESTRUCTIVE = re.compile(
    r"\bDROP\b|\bDELETE\b|\bTRUNCATE\b|\bGRANT\b|\bREVOKE\b|\bALTER\b", re.IGNORECASE
)
_TRANSFORM_HEAD = re.compile(
    r'CREATE\s+(OR\s+REPLACE\s+)?(VIEW|TABLE|MATERIALIZED\s+VIEW)\s+"?([a-z0-9_]+)"?\s*\.\s*"?[a-z0-9_]+"?',
    re.IGNORECASE,
)


def ddl_kind_for_transformation(kind: str) -> TransformationDdlKind:
    k = kind.lower()
    if k in ("cleanse", "join", "rename"):
        return "table"
    return "view"


def assert_transformation_sql(
    sql: str, warehouse_schema: str, expected_ddl: TransformationDdlKind | None = None
) -> None:
    if not _TRANSFORM_ALLOWED.search(sql):
        raise SqlValidationError(
            "Transformation SQL must start with CREATE [OR REPLACE] VIEW / TABLE / MATERIALIZED VIEW."
        )
    if _TRANSFORM_DESTRUCTIVE.search(sql):
        raise SqlValidationError("Transformation SQL contains a forbidden destructive keyword.")

    head = _TRANSFORM_HEAD.search(sql)
    if not head:
        raise SqlValidationError(
            f'Transformation SQL must use a fully-qualified target like "{warehouse_schema}"."my_view".'
        )
    kind_token = head.group(2).upper()
    ddl: TransformationDdlKind = (
        "view" if (kind_token.startswith("VIEW") or "MATERIALIZED" in kind_token) else "table"
    )
    if head.group(3).lower() != warehouse_schema.lower():
        raise SqlValidationError(
            f'Transformation must target schema "{warehouse_schema}", got "{head.group(3)}".'
        )
    if expected_ddl and ddl != expected_ddl:
        raise SqlValidationError(
            f"This transformation should produce a {expected_ddl.upper()} (per its kind), "
            f"but the SQL creates a {ddl.upper()}."
        )


_NORMALIZE_HEAD = re.compile(
    r"^(\s*)CREATE\s+(OR\s+REPLACE\s+)?(VIEW|TABLE|MATERIALIZED\s+VIEW)(\s+)", re.IGNORECASE
)


def normalize_transformation_ddl(sql: str, expected_ddl: TransformationDdlKind) -> str:
    match = _NORMALIZE_HEAD.search(sql)
    if not match:
        return sql
    leading = match.group(1)
    kind = match.group(3)
    current_ddl: TransformationDdlKind = "view" if "VIEW" in kind.upper() else "table"
    if current_ddl == expected_ddl:
        return sql

    replacement = (
        f"{leading}CREATE OR REPLACE VIEW " if expected_ddl == "view" else f"{leading}CREATE TABLE "
    )
    rewritten = sql.replace(match.group(0), replacement, 1)

    if not re.search(r"\bAS\s+SELECT", rewritten, re.IGNORECASE) and re.search(
        r"\bSELECT\b", rewritten, re.IGNORECASE
    ):
        rewritten = re.sub(r"\)\s*SELECT", ") AS SELECT", rewritten, count=1, flags=re.IGNORECASE)
        rewritten = re.sub(r'"\s+SELECT', '" AS SELECT', rewritten, count=1, flags=re.IGNORECASE)
    return rewritten


_FORBIDDEN_IN_FORMULA = re.compile(
    r"\b(CREATE|ALTER|DROP|INSERT|UPDATE|DELETE|TRUNCATE|GRANT|REVOKE|MERGE|COPY|VACUUM|ANALYZE|BEGIN|COMMIT|ROLLBACK)\b",
    re.IGNORECASE,
)
_LOOKS_LIKE_EXPR = re.compile(r'[A-Z_]+\(|"\w+"|\w+\.\w+', re.IGNORECASE)


def assert_measure_formula(formula: str) -> None:
    if not formula or not isinstance(formula, str):
        raise SqlValidationError("sqlFormula is required.")
    if ";" in formula:
        raise SqlValidationError(
            "sqlFormula must be an expression, not a statement. Semicolons are forbidden."
        )
    if _FORBIDDEN_IN_FORMULA.search(formula):
        raise SqlValidationError(
            "sqlFormula contains a forbidden statement keyword (CREATE/ALTER/INSERT/...). "
            "Measures are runtime expressions, not statements."
        )
    if not _LOOKS_LIKE_EXPR.search(formula):
        raise SqlValidationError("sqlFormula does not look like a SQL expression.")
